Service crashes after 6th client connected (panic: runtime error: index out of range)

```
C:\Program Files\MysteriumVPN\bin> .\myst.exe --testnet service openvpn --agreed-terms-and-conditions
2019-03-21T15:45:46.0825054 [Info] Starting Mysterium Node (0.0.0-dev)
2019-03-21T15:45:46.1005195 [Info] [Openvpn check] OpenVPN 2.4.6 x86_64-w64-mingw32 [SSL (OpenSSL)] [LZO] [LZ4] [PKCS11] [AEAD] built on May 15 2018
2019-03-21T15:45:46.1005195 [Info] [Openvpn check] library versions: OpenSSL 1.1.0h  27 Mar 2018, LZO 2.10
2019-03-21T15:45:46.1005195 [Info] [Openvpn check] Windows version 6.2 (Windows 8 or greater) 64bit
2019-03-21T15:45:46.1005195 [Info] [Openvpn check] Originally developed by James Yonan
2019-03-21T15:45:46.1005195 [Info] [Openvpn check] Copyright (C) 2002-2018 OpenVPN Inc <sales@openvpn.net>
2019-03-21T15:45:46.1005195 [Info] [Openvpn check] Custom tag: Compile time defines: enable_async_push=no enable_comp_stub=no enable_crypto=yes enable_crypto_ofb_cfb=yes enable_debug=no enable_def_auth=yes enable_dependency_tracking=no enable_dlopen=unknown enable_dlopen_self=unknown enable_dlopen_self_static=unknown enable_fast_install=needless enable_fragment=yes enable_iproute2=no enable_libtool_lock=yes enable_lz4=yes enable_lzo=yes enable_management=yes enable_multihome=yes enable_pam_dlopen=no enable_pedantic=no enable_pf=yes enable_pkcs11=yes enable_plugin_auth_pam=no enable_plugin_down_root=no enable_plugins=no enable_port_share=yes enable_selinux=no enable_server=yes enable_shared=no enable_shared_with_static_runtimes=yes enable_silent_rules=no enable_small=no enable_static=yes enable_strict=no enable_strict_options=no enable_systemd=no enable_werror=no enable_win32_dll=yes enable_x509_alt_username=no with_aix_soname=aix with_crypto_library=openssl with_gnu_ld=yes with_mem_check=no with_special_build=standalone-2.4.6-2-62 with_sysroot=no
...
...


panic: runtime error: index out of range

goroutine 148 [running]:
github.com/mysteriumnetwork/node/session.(*StorageMemory).Find(0xc0002ad200, 0xc00001afd5, 0x24, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc000ad3200)
        /ext-go/1/src/github.com/mysteriumnetwork/node/session/storage_memory.go:65 +0x172
github.com/mysteriumnetwork/node/services/openvpn/session.(*clientMap).FindClientSession(0xc000b88040, 0x6, 0xc00001afd5, 0x24, 0x0, 0x0, 0x0, 0x0, 0x0, 0x5, ...)
        /ext-go/1/src/github.com/mysteriumnetwork/node/services/openvpn/session/client_map.go:44 +0x9e
github.com/mysteriumnetwork/node/services/openvpn/session.(*Validator).Validate(0xc000b88020, 0x6, 0xc00001afd5, 0x24, 0xc0002c6a95, 0x58, 0x3, 0xf72860, 0xc000b2f8c0)
        /ext-go/1/src/github.com/mysteriumnetwork/node/services/openvpn/session/validator.go:52 +0x90
github.com/mysteriumnetwork/node/services/openvpn/session.(*Validator).Validate-fm(0x6, 0xc00001afd5, 0x24, 0xc0002c6a95, 0x58, 0x90000000064f298, 0x40d479, 0x0)
        /ext-go/1/src/github.com/mysteriumnetwork/node/services/openvpn/service/factory.go:83 +0x63
github.com/mysteriumnetwork/node/vendor/github.com/mysteriumnetwork/go-openvpn/openvpn/middlewares/server/auth.(*middleware).authenticateClient(0xc000ab9590, 0x6, 0x2, 0xc00001afd5, 0x24, 0xc0002c6a95, 0x58, 0x6, 0x0)
        /ext-go/1/src/github.com/mysteriumnetwork/node/vendor/github.com/mysteriumnetwork/go-openvpn/openvpn/middlewares/server/auth/middleware.go:177 +0x265
github.com/mysteriumnetwork/node/vendor/github.com/mysteriumnetwork/go-openvpn/openvpn/middlewares/server/auth.(*middleware).handleClientEvent(0xc000ab9590, 0xc00001ce28, 0x6, 0x6, 0x2, 0xc000142780)
        /ext-go/1/src/github.com/mysteriumnetwork/node/vendor/github.com/mysteriumnetwork/go-openvpn/openvpn/middlewares/server/auth/middleware.go:156 +0x141
github.com/mysteriumnetwork/node/vendor/github.com/mysteriumnetwork/go-openvpn/openvpn/middlewares/server/auth.(*middleware).endOfEvent(0xc000ab9590)
        /ext-go/1/src/github.com/mysteriumnetwork/node/vendor/github.com/mysteriumnetwork/go-openvpn/openvpn/middlewares/server/auth/middleware.go:143 +0x51
github.com/mysteriumnetwork/node/vendor/github.com/mysteriumnetwork/go-openvpn/openvpn/middlewares/server/auth.(*middleware).ConsumeLine(0xc000ab9590, 0xc000ade020, 0xf, 0xc000b731b0, 0x0, 0x0)
        /ext-go/1/src/github.com/mysteriumnetwork/node/vendor/github.com/mysteriumnetwork/go-openvpn/openvpn/middlewares/server/auth/middleware.go:108 +0x5a8
github.com/mysteriumnetwork/node/vendor/github.com/mysteriumnetwork/go-openvpn/openvpn/management.(*Management).deliverOpenvpnManagementEvents(0xc0001bebd0, 0xc000052420)
        /ext-go/1/src/github.com/mysteriumnetwork/node/vendor/github.com/mysteriumnetwork/go-openvpn/openvpn/management/listener.go:226 +0x323
github.com/mysteriumnetwork/node/vendor/github.com/mysteriumnetwork/go-openvpn/openvpn/management.(*Management).serveNewConnection.func2(0xc00014e010, 0xc0001bebd0, 0xc000052420)
        /ext-go/1/src/github.com/mysteriumnetwork/node/vendor/github.com/mysteriumnetwork/go-openvpn/openvpn/management/listener.go:167 +0x62
created by github.com/mysteriumnetwork/node/vendor/github.com/mysteriumnetwork/go-openvpn/openvpn/management.(*Management).serveNewConnection
        /ext-go/1/src/github.com/mysteriumnetwork/node/vendor/github.com/mysteriumnetwork/go-openvpn/openvpn/management/listener.go:165 +0x2ea
```

--- a/session/dto.go
+++ b/session/dto.go
@@ -17,7 +17,11 @@
 
 package session
 
-import "github.com/mysteriumnetwork/node/identity"
+import (
+	"time"
+
+	"github.com/mysteriumnetwork/node/identity"
+)
 
 // ID represents session id type
 type ID string
@@ -34,6 +38,7 @@ type Session struct {
 	ConsumerID identity.Identity
 	Done       chan struct{}
 	Config     ServiceConfiguration
+	CreatedAt  time.Time
 	Last       bool
 }
 

--- a/session/manager.go
+++ b/session/manager.go
@@ -21,6 +21,7 @@ import (
 	"encoding/json"
 	"errors"
 	"sync"
+	"time"
 
 	log "github.com/cihub/seelog"
 	"github.com/mysteriumnetwork/node/identity"
@@ -130,6 +131,7 @@ func (manager *Manager) Create(consumerID identity.Identity, issuerID identity.I
 	sessionInstance.ConsumerID = consumerID
 	sessionInstance.Done = make(chan struct{})
 	sessionInstance.Config = config
+	sessionInstance.CreatedAt = time.Now().UTC()
 
 	balanceTracker, err := manager.balanceTrackerFactory(consumerID, identity.FromAddress(manager.currentProposal.ProviderID), issuerID)
 	if err != nil {

--- a/session/manager_test.go
+++ b/session/manager_test.go
@@ -73,7 +73,13 @@ func TestManager_Create_StoresSession(t *testing.T) {
 	sessionInstance, err := manager.Create(consumerID, consumerID, currentProposalID, nil, requestConfig)
 	expectedResult.Done = sessionInstance.Done
 	assert.NoError(t, err)
-	assert.Exactly(t, expectedResult, sessionInstance)
+
+	assert.Equal(t, expectedResult.Config, sessionInstance.Config)
+	assert.Equal(t, expectedResult.Last, sessionInstance.Last)
+	assert.Equal(t, expectedResult.Done, sessionInstance.Done)
+	assert.Equal(t, expectedResult.ConsumerID, sessionInstance.ConsumerID)
+	assert.Equal(t, expectedResult.ID, sessionInstance.ID)
+	assert.False(t, sessionInstance.CreatedAt.IsZero())
 }
 
 func TestManager_Create_RejectsUnknownProposal(t *testing.T) {

--- a/session/storage_memory.go
+++ b/session/storage_memory.go
@@ -18,23 +18,22 @@
 package session
 
 import (
+	"sort"
 	"sync"
 )
 
 // NewStorageMemory initiates new session storage
 func NewStorageMemory() *StorageMemory {
 	return &StorageMemory{
-		sessions:  make([]Session, 0),
-		idToIndex: make(map[ID]int),
-		lock:      sync.Mutex{},
+		sessions: make(map[ID]Session),
+		lock:     sync.Mutex{},
 	}
 }
 
 // StorageMemory maintains all currents sessions in memory
 type StorageMemory struct {
-	sessions  []Session
-	idToIndex map[ID]int
-	lock      sync.Mutex
+	sessions map[ID]Session
+	lock     sync.Mutex
 }
 
 // Add puts given session to storage. Multiple sessions per peerID is possible in case different services are used
@@ -42,18 +41,26 @@ func (storage *StorageMemory) Add(sessionInstance Session) {
 	storage.lock.Lock()
 	defer storage.lock.Unlock()
 
-	sessionIndex := len(storage.sessions)
-	storage.sessions = append(storage.sessions, sessionInstance)
-
-	storage.idToIndex[sessionInstance.ID] = sessionIndex
+	storage.sessions[sessionInstance.ID] = sessionInstance
 }
 
 // GetAll returns all sessions in storage
-func (storage *StorageMemory) GetAll() ([]Session, error) {
+func (storage *StorageMemory) GetAll() []Session {
 	storage.lock.Lock()
 	defer storage.lock.Unlock()
 
-	return storage.sessions, nil
+	// we're never gonna have more than 100000 sessions ongoing on a single node - performance here should not be an issue.
+	// see Benchmark_Storage_GetAll
+	sessions := make([]Session, len(storage.sessions))
+
+	i := 0
+	for _, value := range storage.sessions {
+		sessions[i] = value
+		i++
+	}
+
+	sort.Slice(sessions, func(i, j int) bool { return sessions[i].CreatedAt.Before(sessions[j].CreatedAt) })
+	return sessions
 }
 
 // Find returns underlying session instance
@@ -61,8 +68,7 @@ func (storage *StorageMemory) Find(id ID) (Session, bool) {
 	storage.lock.Lock()
 	defer storage.lock.Unlock()
 
-	if sessionIndex, found := storage.idToIndex[id]; found {
-		instance := storage.sessions[sessionIndex]
+	if instance, found := storage.sessions[id]; found {
 		if len(storage.sessions) == 1 {
 			instance.Last = true
 		}
@@ -77,8 +83,7 @@ func (storage *StorageMemory) Remove(id ID) {
 	storage.lock.Lock()
 	defer storage.lock.Unlock()
 
-	if sessionIndex, found := storage.idToIndex[id]; found {
-		delete(storage.idToIndex, id)
-		storage.sessions = append(storage.sessions[:sessionIndex], storage.sessions[sessionIndex+1:]...)
+	if _, found := storage.sessions[id]; found {
+		delete(storage.sessions, id)
 	}
 }

--- a/session/storage_memory.go
+++ b/session/storage_memory.go
@@ -18,7 +18,6 @@
 package session
 
 import (
-	"sort"
 	"sync"
 )
 
@@ -58,8 +57,6 @@ func (storage *StorageMemory) GetAll() []Session {
 		sessions[i] = value
 		i++
 	}
-
-	sort.Slice(sessions, func(i, j int) bool { return sessions[i].CreatedAt.Before(sessions[j].CreatedAt) })
 	return sessions
 }
 

--- a/session/storage_memory.go
+++ b/session/storage_memory.go
@@ -79,8 +79,5 @@ func (storage *StorageMemory) Find(id ID) (Session, bool) {
 func (storage *StorageMemory) Remove(id ID) {
 	storage.lock.Lock()
 	defer storage.lock.Unlock()
-
-	if _, found := storage.sessions[id]; found {
-		delete(storage.sessions, id)
-	}
+	delete(storage.sessions, id)
 }

--- a/session/storage_memory_test.go
+++ b/session/storage_memory_test.go
@@ -18,7 +18,9 @@
 package session
 
 import (
+	"fmt"
 	"testing"
+	"time"
 
 	"github.com/stretchr/testify/assert"
 )
@@ -56,36 +58,29 @@ func TestStorage_Add(t *testing.T) {
 	storage.Add(sessionNew)
 	assert.Exactly(
 		t,
-		[]Session{sessionExisting, sessionNew},
+		map[ID]Session{sessionExisting.ID: sessionExisting, sessionNew.ID: sessionNew},
 		storage.sessions,
-	)
-	assert.Exactly(
-		t,
-		map[ID]int{
-			sessionExisting.ID: 0,
-			sessionNew.ID:      1,
-		},
-		storage.idToIndex,
 	)
 }
 
 func TestStorage_GetAll(t *testing.T) {
-	sessionFirst := Session{ID: ID("id1")}
-	sessionSecond := Session{ID: ID("id2")}
+	sessionFirst := Session{
+		ID: ID("id1"),
+	}
+	sessionSecond := Session{
+		ID:        ID("id2"),
+		CreatedAt: time.Now(),
+	}
+
 	storage := &StorageMemory{
-		sessions: []Session{
-			sessionFirst,
-			sessionSecond,
-		},
-		idToIndex: map[ID]int{
-			sessionFirst.ID:  0,
-			sessionSecond.ID: 1,
+		sessions: map[ID]Session{
+			sessionFirst.ID:  sessionFirst,
+			sessionSecond.ID: sessionSecond,
 		},
 	}
 
-	sessions, err := storage.GetAll()
-	assert.NoError(t, err)
-	assert.Equal(t, sessions, []Session{sessionFirst, sessionSecond})
+	sessions := storage.GetAll()
+	assert.Equal(t, []Session{sessionFirst, sessionSecond}, sessions)
 }
 
 func TestStorage_Remove(t *testing.T) {
@@ -93,14 +88,41 @@ func TestStorage_Remove(t *testing.T) {
 
 	storage.Remove(sessionExisting.ID)
 	assert.Len(t, storage.sessions, 0)
-	assert.Len(t, storage.idToIndex, 0)
+}
+
+func TestStorage_Remove_Does_Not_Panic(t *testing.T) {
+	id4 := ID("id4")
+	storage := mockStorage(sessionExisting)
+	sessionFirst := Session{ID: id4}
+	sessionSecond := Session{ID: ID("id3")}
+	storage.Add(sessionFirst)
+	storage.Add(sessionSecond)
+	storage.Remove(id4)
+	storage.Remove(ID("id3"))
+	assert.Len(t, storage.sessions, 1)
 }
 
 func mockStorage(sessionInstance Session) *StorageMemory {
 	return &StorageMemory{
-		sessions: []Session{sessionInstance},
-		idToIndex: map[ID]int{
-			sessionInstance.ID: 0,
-		},
+		sessions: map[ID]Session{sessionInstance.ID: sessionInstance},
 	}
+}
+
+// to avoid compiler optimizing away our bench
+var benchmarkStorageGetAllResult int
+
+func Benchmark_Storage_GetAll(b *testing.B) {
+	// Findings are as follows - with 100k sessions, we should be fine with a performance of 0.04s on my mac
+	storage := NewStorageMemory()
+	sessionsToStore := 100000
+	for i := 0; i < sessionsToStore; i++ {
+		storage.Add(Session{ID: ID(fmt.Sprintf("ID%v", i)), CreatedAt: time.Now()})
+	}
+
+	var r int
+	for n := 0; n < b.N; n++ {
+		storedValues := storage.GetAll()
+		r += len(storedValues)
+	}
+	benchmarkStorageGetAllResult = r
 }

--- a/session/storage_memory_test.go
+++ b/session/storage_memory_test.go
@@ -80,7 +80,8 @@ func TestStorage_GetAll(t *testing.T) {
 	}
 
 	sessions := storage.GetAll()
-	assert.Equal(t, []Session{sessionFirst, sessionSecond}, sessions)
+	assert.Contains(t, sessions, sessionFirst)
+	assert.Contains(t, sessions, sessionSecond)
 }
 
 func TestStorage_Remove(t *testing.T) {

--- a/session/storage_memory_test.go
+++ b/session/storage_memory_test.go
@@ -91,6 +91,14 @@ func TestStorage_Remove(t *testing.T) {
 	assert.Len(t, storage.sessions, 0)
 }
 
+func TestStorage_RemoveLast(t *testing.T) {
+	storage := &StorageMemory{
+		sessions: map[ID]Session{},
+	}
+	storage.Remove(sessionExisting.ID)
+	assert.Len(t, storage.sessions, 0)
+}
+
 func TestStorage_Remove_Does_Not_Panic(t *testing.T) {
 	id4 := ID("id4")
 	storage := mockStorage(sessionExisting)

--- a/session/storage_memory_test.go
+++ b/session/storage_memory_test.go
@@ -91,7 +91,7 @@ func TestStorage_Remove(t *testing.T) {
 	assert.Len(t, storage.sessions, 0)
 }
 
-func TestStorage_RemoveLast(t *testing.T) {
+func TestStorage_RemoveNonExisting(t *testing.T) {
 	storage := &StorageMemory{
 		sessions: map[ID]Session{},
 	}

--- a/tequilapi/endpoints/service_sessions.go
+++ b/tequilapi/endpoints/service_sessions.go
@@ -66,10 +66,6 @@ func NewServiceSessionsEndpoint(sessionStorage serviceSessionStorage) *serviceSe
 //     description: List of sessions
 //     schema:
 //       "$ref": "#/definitions/ServiceSessionListDTO"
-//   500:
-//     description: Internal server error
-//     schema:
-//       "$ref": "#/definitions/ErrorMessageDTO"
 func (endpoint *serviceSessionsEndpoint) List(resp http.ResponseWriter, request *http.Request, params httprouter.Params) {
 	sessions := endpoint.sessionStorage.GetAll()
 

--- a/tequilapi/endpoints/service_sessions.go
+++ b/tequilapi/endpoints/service_sessions.go
@@ -19,6 +19,7 @@ package endpoints
 
 import (
 	"net/http"
+	"sort"
 
 	"github.com/julienschmidt/httprouter"
 	"github.com/mysteriumnetwork/node/session"
@@ -71,6 +72,8 @@ func NewServiceSessionsEndpoint(sessionStorage serviceSessionStorage) *serviceSe
 //       "$ref": "#/definitions/ErrorMessageDTO"
 func (endpoint *serviceSessionsEndpoint) List(resp http.ResponseWriter, request *http.Request, params httprouter.Params) {
 	sessions := endpoint.sessionStorage.GetAll()
+
+	sort.Slice(sessions, func(i, j int) bool { return sessions[i].CreatedAt.Before(sessions[j].CreatedAt) })
 
 	sessionsSerializable := serviceSessionsList{
 		Sessions: mapServiceSessions(sessions, serviceSessionToDto),

--- a/tequilapi/endpoints/service_sessions.go
+++ b/tequilapi/endpoints/service_sessions.go
@@ -42,7 +42,7 @@ type serviceSession struct {
 }
 
 type serviceSessionStorage interface {
-	GetAll() ([]session.Session, error)
+	GetAll() []session.Session
 }
 
 type serviceSessionsEndpoint struct {
@@ -70,11 +70,7 @@ func NewServiceSessionsEndpoint(sessionStorage serviceSessionStorage) *serviceSe
 //     schema:
 //       "$ref": "#/definitions/ErrorMessageDTO"
 func (endpoint *serviceSessionsEndpoint) List(resp http.ResponseWriter, request *http.Request, params httprouter.Params) {
-	sessions, err := endpoint.sessionStorage.GetAll()
-	if err != nil {
-		utils.SendError(resp, err, http.StatusInternalServerError)
-		return
-	}
+	sessions := endpoint.sessionStorage.GetAll()
 
 	sessionsSerializable := serviceSessionsList{
 		Sessions: mapServiceSessions(sessions, serviceSessionToDto),

--- a/tequilapi/endpoints/service_sessions_test.go
+++ b/tequilapi/endpoints/service_sessions_test.go
@@ -22,6 +22,7 @@ import (
 	"net/http"
 	"net/http/httptest"
 	"testing"
+	"time"
 
 	"github.com/stretchr/testify/assert"
 
@@ -52,9 +53,16 @@ func Test_ServiceSessionsEndpoint_List(t *testing.T) {
 	)
 	assert.Nil(t, err)
 
+	anotherSession := session.Session{
+		ID:         session.ID("session2"),
+		ConsumerID: identity.FromAddress("consumer1"),
+		CreatedAt:  time.Now(),
+	}
+
 	ssm := &serviceSessionStorageMock{
 		sessionsToReturn: []session.Session{
 			serviceSessionMock,
+			anotherSession,
 		},
 	}
 
@@ -66,6 +74,7 @@ func Test_ServiceSessionsEndpoint_List(t *testing.T) {
 	err = json.Unmarshal(resp.Body.Bytes(), parsedResponse)
 	assert.Nil(t, err)
 	assert.EqualValues(t, serviceSessionToDto(serviceSessionMock), parsedResponse.Sessions[0])
+	assert.EqualValues(t, serviceSessionToDto(anotherSession), parsedResponse.Sessions[1])
 }
 
 type serviceSessionStorageMock struct {

--- a/tequilapi/endpoints/service_sessions_test.go
+++ b/tequilapi/endpoints/service_sessions_test.go
@@ -19,8 +19,6 @@ package endpoints
 
 import (
 	"encoding/json"
-	"errors"
-	"fmt"
 	"net/http"
 	"net/http/httptest"
 	"testing"
@@ -55,7 +53,6 @@ func Test_ServiceSessionsEndpoint_List(t *testing.T) {
 	assert.Nil(t, err)
 
 	ssm := &serviceSessionStorageMock{
-		errToReturn: nil,
 		sessionsToReturn: []session.Session{
 			serviceSessionMock,
 		},
@@ -71,36 +68,10 @@ func Test_ServiceSessionsEndpoint_List(t *testing.T) {
 	assert.EqualValues(t, serviceSessionToDto(serviceSessionMock), parsedResponse.Sessions[0])
 }
 
-func Test_ServiceSessionsEndpoint_ListBubblesError(t *testing.T) {
-	req, err := http.NewRequest(
-		http.MethodGet,
-		"/irrelevant",
-		nil,
-	)
-	assert.Nil(t, err)
-
-	mockErr := errors.New("something exploded")
-	ssm := &serviceSessionStorageMock{
-		errToReturn:      mockErr,
-		sessionsToReturn: []session.Session{},
-	}
-
-	resp := httptest.NewRecorder()
-	handlerFunc := NewServiceSessionsEndpoint(ssm).List
-	handlerFunc(resp, req, nil)
-
-	assert.Equal(t, http.StatusInternalServerError, resp.Code)
-	assert.Equal(t,
-		fmt.Sprintf(`{"message":%q}%v`, mockErr.Error(), "\n"),
-		resp.Body.String(),
-	)
-}
-
 type serviceSessionStorageMock struct {
 	sessionsToReturn []session.Session
-	errToReturn      error
 }
 
-func (ssm *serviceSessionStorageMock) GetAll() ([]session.Session, error) {
-	return ssm.sessionsToReturn, ssm.errToReturn
+func (ssm *serviceSessionStorageMock) GetAll() []session.Session {
+	return ssm.sessionsToReturn
 }

